CMakeLists.txt

```diff
@@ -1,32 +1,79 @@
 cmake_minimum_required(VERSION 3.1)
 
-
-set(CMAKE_DISABLE_SOURCE_CHANGES ON)
-set(CMAKE_DISABLE_IN_SOURCE_BUILD ON)
-
 project(NetKet)
 
-set (CMAKE_CXX_STANDARD 11)
-set (NETKET_SOURCE_DIR ${CMAKE_CURRENT_SOURCE_DIR}/NetKet/)
-set (EXTERNAL_INCLUDE_DIR ${CMAKE_CURRENT_SOURCE_DIR}/External/)
+include(CTest) # adds option BUILD_TESTING (default ON)
+include(ExternalProject)
 
+set(CMAKE_CXX_STANDARD 11)
+set(CMAKE_CXX_EXTENSIONS OFF)
 
-set (EIGEN3_INCLUDE_DIR "${EXTERNAL_INCLUDE_DIR}" CACHE STRING "Eigen3 Include Directory" )
-set (JSON_INCLUDE_DIR "${EXTERNAL_INCLUDE_DIR}/Json/" CACHE STRING "JSON for Modern C++ Include Directory" )
+#
+# Dependencies
+#
 
-#Find MPI
+# json
+################################################################################
+if (NOT EXISTS "${CMAKE_BINARY_DIR}/External/json/json.hpp")
+    file(DOWNLOAD
+        "https://github.com/nlohmann/json/releases/download/v3.2.0/json.hpp"
+        "External/json/json.hpp")
+endif()
+add_library(nlohmann_json INTERFACE)
+target_include_directories(nlohmann_json
+    INTERFACE "${CMAKE_BINARY_DIR}/External/json")
+
+# optional-lite
+################################################################################
+if (NOT EXISTS "${CMAKE_BINARY_DIR}/External/optional-lite/nonstd/optional.hpp")
+    file(DOWNLOAD
+        "https://github.com/martinmoene/optional-lite/releases/download/v3.0.0/optional.hpp"
+        "External/optional-lite/nonstd/optional.hpp")
+endif()
+add_library(optional_lite INTERFACE)
+target_include_directories(optional_lite
+    INTERFACE "${CMAKE_BINARY_DIR}/External/optional-lite")
+
+# Eigen3
+################################################################################
+ExternalProject_Add(
+    eigen_project
+    SOURCE_DIR "${CMAKE_BINARY_DIR}/External/Eigen3"
+    URL "https://github.com/eigenteam/eigen-git-mirror/archive/3.3.5.tar.gz"
+    URL_HASH MD5=305eed19157335016bde9b4d57ee3d35
+    CONFIGURE_COMMAND ""
+    BUILD_COMMAND ""
+    INSTALL_COMMAND ""
+)
+ExternalProject_Get_Property(eigen_project SOURCE_DIR)
+add_library(Eigen3 INTERFACE)
+target_include_directories(Eigen3 SYSTEM INTERFACE ${SOURCE_DIR})
+
+# MPI
+################################################################################
 find_package(MPI REQUIRED)
-if (MPI_FOUND)
-    include_directories(SYSTEM ${MPI_INCLUDE_PATH})
-else (MPI_FOUND)
-    message(SEND_ERROR "This application cannot compile without MPI")
-endif (MPI_FOUND)
 
-include_directories(${MPI_INCLUDE_PATH})
```

Comment:
I think that's why tests are failing on some compilers. It seems you need to restore the MPI_INCLUDE_PATH

Reply:
71fe221 should fix the failures